refactor: have CCoins* data managed under CChainState

jamesob opened the pull request:
This is part of the [assumeutxo project](https://github.com/bitcoin/bitcoin/projects/11):

Parent PR: #15606 
Issue: #15605 
Specification: https://github.com/jamesob/assumeutxo-docs/tree/2019-04-proposal/proposal

---

This change encapsulates UTXO set data within CChainState instances, removing global data `pcoinsTip` and `pcoinsviewdb`. This is necessary if we want to maintain multiple chainstates with their own rendering of the UTXO set.

We introduce a class CoinsViews which consolidates the construction of a CCoins* hierarchy.

This commit could be broken into smaller pieces, but it would require more ephemeral diffs to, e.g., temporarily change CCoinsViewDB's constructor invocations.

MarcoFalke commented on src/validation.h:
```diff
@@ -19,6 +19,8 @@
 #include <sync.h>
 #include <txmempool.h> // For CTxMemPool::cs
+#include <txdb.h>
+#include <util/system.h>
```
I'd prefer to make the call to `GetDataDir` in the cpp file.

MarcoFalke commented on src/validation.h:
```diff
@@ -659,11 +718,7 @@ CChain& ChainActive();
-/** Global variable that points to the coins database (protected by cs_main) */
-extern std::unique_ptr<CCoinsViewDB> pcoinsdbview;
-
-/** Global variable that points to the active CCoinsView (protected by cs_main) */
```
Why are you removing the comment that this data is protected by cs_main?

jamesob replied:
Good point - since it was easy to do, I've tacked on a commit that adds lock annotations to the equivalent CoinsViews members.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -3853,12 +3873,13 @@ bool static LoadBlockIndexDB(const CChainParams& chainparams) EXCLUSIVE_LOCKS_RE
 {
     AssertLockHeld(cs_main);
-    assert(!pcoinsTip->GetBestBlock().IsNull()); // Never called when the coins view is empty
+    CCoinsViewCache* coins_cache = &::ChainstateActive().CoinsTip();
```
```suggestion
    const CCoinsViewCache& coins_cache = ::ChainstateActive().CoinsTip();
```

MarcoFalke commented on src/validation.cpp:
```diff
@@ -81,11 +81,17 @@ namespace {
-CChainState& ChainstateActive() { return g_chainstate; }
+CChainState& ChainstateActive() {
+    assert(g_chainstate);
+    return *g_chainstate.get();
```
```suggestion
    return *g_chainstate;
```

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -1291,11 +1291,12 @@ bool static AlreadyHave(const CInv& inv) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
                 LOCK(g_cs_orphans);
                 if (mapOrphanTransactions.count(inv.hash)) return true;
             }
+            CCoinsViewCache& coins_cache = ::ChainstateActive().CoinsTip();
```
```suggestion
            const CCoinsViewCache& coins_cache = ::ChainstateActive().CoinsTip();
```

maflcko reviewed (commented):
Would it make sense to split this into dumb refactoring and then hard-to-review changes, like the on in init.cpp?

An early commit (scripted-diff or not) could add `ChainstateActive::CoinsTip`, a dumb alias to `return pcoinsTip;` and replace the usages. The later commit could then make the other changes in this pull.

jamesob committed:
refactor: pcoinsTip -> CChainState::CoinsTip()

This aliasing makes subsequent commits easier to review; eventually CoinsTip()
will return the CCoinsViewCache managed by CChainState.

jamesob committed:
refactor: have CCoins* data managed under CChainState

This change encapsulates UTXO set data within CChainState instances, removing
global data `pcoinsTip` and `pcoinsviewdb`. This is necessary if we want to
maintain multiple chainstates with their own rendering of the UTXO set.

We introduce a class CoinsViews which consolidates the construction of a
CCoins* hierarchy. Construction of its various pieces (db, coinscatcher,
in-memory cache) is split up so that we avoid flushing bad state to disk if
startup is interrupted.

We also introduce `CChainState::CanFlushToDisk()` which tells us when it is
safe to flush the chainstate based on this partial construction.

This commit could be broken into smaller pieces, but it would require more
ephemeral diffs to, e.g., temporarily change CCoinsViewDB's constructor
invocations.

Other changes:

- A parameter has been added to the CCoinsViewDB constructor that allows the
  name of the corresponding leveldb directory to be specified.

Thanks to Russell Yanofsky and Marco Falke for helpful feedback.

jamesob committed:
Cover UTXO set access with lock annotations

i.e. any CoinsViews members. Adds a lock acquisition to `gettxoutsetinfo` RPC
to comply with added annotations.

Co-authored-by: Russell Yanofsky <russ@yanofsky.org>

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -1062,6 +1062,8 @@ static UniValue gettxoutsetinfo(const JSONRPCRequest& request)
     CCoinsStats stats;
     ::ChainstateActive().ForceFlushStateToDisk();
+
+    LOCK(::cs_main);
```
In commit "Cover UTXO set access with lock annotations" (ceca2c92b79a8a2c5ae37d03e99de80ec55895c3)

Commit message is a bit misleading imo, because this is adding a new lock, not just new annotations.

But more importantly, GetUTXOStats is acquiring and releasing cs_main internally, so this change seems like it is complicating locking and also locking too much. Maybe a change like the following would make sense?

```diff
-    LOCK(::cs_main);
-    if (GetUTXOStats(&::ChainstateActive().CoinsDB(), stats)) {
+    CCoinsView* coins_view = WITH_LOCK(cs_main, return &ChainstateActive().CoinsDB());
+    if (GetUTXOStats(coins_view, stats)) {
```

jamesob replied:
Ah that's a good point - I forgot that `GetUTXOStats` only briefly acquires cs_main to get the height for a certain blockhash; it (counterintuitively) doesn't need to hold cs_main to iterate through the coins db because leveldb's implicit use of snapshots with cursors. 

I think the `WITH_LOCK` usage above obeys the letter of the law but not the spirit since technically the coinsdb could be modified in between that line and the GetUTXOStats call, but in this context I don't think it matters. Thanks for the suggestion!

ryanofsky commented on src/validation.h:
```diff
@@ -686,8 +714,7 @@ CChain& ChainActive();
 
-/** Global variable that points to the coins database (protected by cs_main) */
-extern std::unique_ptr<CCoinsViewDB> pcoinsdbview;
+extern std::unique_ptr<CChainState> g_chainstate;
```
In commit "refactor: have CCoins* data managed under CChainState" (eed4e62ad6b162ff0d3e2b87eb8762a689678bca)

Can you add a comment here about when it's appropriate to access `g_chainstate` variable directly, as opposed to calling `ChainstateActive()`? I see code in `init.cpp` is using both, but mostly the former, which is different than what I'd expect. I'd expect init code to just create a CChainState object and assign it it to `g_chainstate` once rather than accessing the `g_chainstate` global repeatedly.

ryanofsky commented on src/validation.cpp:
```diff
@@ -1050,6 +1054,21 @@ CoinsViews::CoinsViews(
+    // in a future commit.
+    ) : m_blockman(g_blockman)
+{
+    m_coins_views.reset(new CoinsViews(
```
In commit "refactor: have CCoins* data managed under CChainState" (eed4e62ad6b162ff0d3e2b87eb8762a689678bca)

Suggest `x = MakeUnique<X>()` instead of `x.reset(new X())` to avoid raw pointers and new in C++, which are suspicious.

ryanofsky commented on src/init.cpp:
```diff
@@ -1464,10 +1460,18 @@ bool AppInitMain(InitInterfaces& interfaces)
             bool is_coinsview_empty;
             try {
                 LOCK(cs_main);
+                g_chainstate.reset(new CChainState(
```
In commit "refactor: have CCoins* data managed under CChainState" (eed4e62ad6b162ff0d3e2b87eb8762a689678bca)

Suggest `x = MakeUnique<X>()` instead of `x.reset(new X())` to avoid raw pointers and new in C++, which are suspicious.

ryanofsky commented on src/test/setup_common.cpp:
```diff
@@ -82,8 +82,8 @@ TestingSetup::TestingSetup(const std::string& chainName) : BasicTestingSetup(cha
     pblocktree.reset(new CBlockTreeDB(1 << 20, true));
-    pcoinsdbview.reset(new CCoinsViewDB(1 << 23, true));
-    pcoinsTip.reset(new CCoinsViewCache(pcoinsdbview.get()));
+    g_chainstate.reset(new CChainState(
```
In commit "refactor: have CCoins* data managed under CChainState" (eed4e62ad6b162ff0d3e2b87eb8762a689678bca)

Suggest `x = MakeUnique<X>()` instead of `x.reset(new X())` to avoid raw pointers and new in C++, which are suspicious.

ryanofsky commented on src/init.cpp:
```diff
@@ -1518,33 +1522,23 @@ bool AppInitMain(InitInterfaces& interfaces)
                 // At this point we're either in reindex or we've loaded a useful
                 // block tree into BlockIndex()!
 
-                pcoinsdbview.reset(new CCoinsViewDB(nCoinDBCache, false, fReset || fReindexChainState));
```
In commit "refactor: have CCoins* data managed under CChainState" (eed4e62ad6b162ff0d3e2b87eb8762a689678bca)

Could you specify in the commit message whether this is changing initialization order in any significant way? It seems like utxo database is being loaded a lot earlier now.

ryanofsky commented on src/txdb.h:
```diff
@@ -48,7 +48,15 @@ class CCoinsViewDB final : public CCoinsView
+    explicit CCoinsViewDB(
+        fs::path ldb_path,
+        size_t nCacheSize,
+        bool fMemory = false,
```
In commit "refactor: have CCoins* data managed under CChainState" (eed4e62ad6b162ff0d3e2b87eb8762a689678bca)

Can we get rid of these default values? It seems like they are not actually used anymore, and default booleans params  in general tend to make code trickier to read and mask bugs.

ryanofsky reviewed (approved):
utACK ceca2c92b79a8a2c5ae37d03e99de80ec55895c3. This looks good and is very straightforward. Feel free to ignore all my suggestions below! (They are not important and possibly not coherent.)

jamesob commented on src/validation.cpp:
```diff
@@ -1038,6 +1038,18 @@ CAmount GetBlockSubsidy(int nHeight, const Consensus::Params& consensusParams)
+    bool in_memory,
+    bool should_wipe)
+{
+    m_dbview.reset(new CCoinsViewDB(
```
Ah, oops - you're right. Shouldn't have tried to split those commits out.

jamesob commented:
Thanks for the feedback, @ryanofsky. I think I've incorporated all of it.

ariard commented on src/validation.h:
```diff
@@ -505,6 +505,9 @@ class BlockManager {
 };
 
+/** Global variable that points to the active CCoinsView (protected by cs_main) */
+extern std::unique_ptr<CCoinsViewCache> pcoinsTip;
```
Did you remove it in last commit ? I still see it at d0e1e37

ariard replied:
Also comment in ATMP refers to it

jamesob replied:
Good catch! Forgot to remove this after I split out the `CoinsTip()` rename commit.

ariard commented on src/validation.h:
```diff
@@ -556,12 +574,21 @@ class CChainState {
+        size_t cache_size_bytes,
+        bool in_memory,
+        bool should_wipe,
+        std::string leveldb_name = "chainstate");
```
Hmmm is there any risk than having a default leveldb_name cause database collisions for different CChainState at some point ?

jamesob replied:
This seems like something that would be caught pretty easily in tests if the confusion ever did happen, and I'm not sure if repeating `"chainstate"` in a few different places inline is any better. That said I'm happy to make this change if others agree it seems worthwhile.

ariard commented on src/validation.h:
```diff
@@ -508,6 +508,24 @@ class BlockManager {
+class CoinsViews {
+
+public:
+    std::unique_ptr<CCoinsViewDB> m_dbview;
```
nit : I know these fields weren't documented as global, but maybe add comment explaining the aforementioned hierarchy between them

ariard reviewed (approved):
ACK d0e1e37 minus `pcoinsTip` removal.

Also I checked the other access to CoinsViews members via CoinsTip like in `src/rpc/blockchain.cpp` or `src/rpc/rawtransaction.cpp`. I've found every to be annotated or with a lock before, but was wandering if we shouldn't rely on `EXCLUSIVE_LOCK_REQUIRED` if possible and not `AssertLockHeld`. What's the exact diff between them ?

sipa commented:
`EXCLUSIVE_LOCK_REQUIRED` is checked at compile time, but only when the compiler has certain flags enabled (only on Clang, I believe).

`AssertLockHeld` is checked at runtime when lock debugging is compiled in (available for all compilers, but only enabled in debug builds).

ariard commented:
Yes `EXCLUSIVE_LOCK_REQUIRED` is defined in `threadsafety.h` and is part of Clang static analysis on Mac OS X. But reviewing last commit, e.g in `ActivateBestChainStep` we only use `AssertLockHeld` but in fact we should use Clang attribute (or maybe both). Is there a consistent policy on which one to favour ?

jamesob commented:
Thanks for the look, @ariard. Feedback pushed.

ryanofsky reviewed (approved):
utACK c860b44ce33f424de7f0fe5f96efa65c58bdb531. There have just been a lot of small changes since last review: rebase for bilingual_str, many new code comments and commit comments, more MakeUniques and c++ references replacing raw pointers, avoiding recursive cs_main lock in gettxoutsetinfo.

re: https://github.com/bitcoin/bitcoin/pull/16443#issuecomment-514864651
> Is there a consistent policy on which one to favour ?

It's good to use EXCLUSIVE_LOCK_REQUIRED wherever possible. AssertLockHeld is older and not used as much recently, but still useful for compilers other than clang and cases where clang annotations don't work (sometimes when there are private mutexes, locks in callbacks and virtual methods).

MarcoFalke commented:
https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#known-limitations

MarcoFalke commented:
In commit 53c8d59d18 refactor: have CCoins* data managed under CChainState
```
   - The initialization order in init.cpp is changed slightly: the coinsdb (leveldb)
      creation is moved to be earlier than block index and genesis block loading,
      though this shouldn't have any noticeable effect.
```
Could you elaborate a bit more why it is save to flush the chainstate when shutdown is requested during the init sequence? As I read the code, previously we wouldn't flush and now we flush an empty db (potentially not having the best block hash set, because genesis is not loaded?).

MarcoFalke commented on src/validation.cpp:
```diff
@@ -2177,9 +2205,10 @@ void static UpdateTip(const CBlockIndex *pindexNew, const CChainParams& chainPar
+    EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
 {
-    CBlockIndex *pindexDelete = m_chain.Tip();
+    CBlockIndex* pindexDelete = m_chain.Tip();
```
Why are you removing the mempool lock annotation? Aren't the lock annotations in the header correct?

jamesob replied:
Ah you're right - dumb mistake on my part.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -2297,7 +2326,8 @@ class ConnectTrace {
  */
-bool CChainState::ConnectTip(CValidationState& state, const CChainParams& chainparams, CBlockIndex* pindexNew, const std::shared_ptr<const CBlock>& pblock, ConnectTrace& connectTrace, DisconnectedBlockTransactions &disconnectpool)
+bool CChainState::ConnectTip(CValidationState& state, const CChainParams& chainparams, CBlockIndex* pindexNew, const std::shared_ptr<const CBlock>& pblock, ConnectTrace& connectTrace, DisconnectedBlockTransactions& disconnectpool)
+    EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
```
Same

maflcko reviewed (commented):
ACK c860b44ce3, code looks good overall. Running into segfaults when testing

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK c860b44ce3, code looks good overall. Running into segfaults when testing
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUi8YAwAnzGBHNruwZp6wvxhnfnew86jtl74PDidtnPwvGX9JiMOeY07UF70JOnO
OwTXN9IV7ZSAQklT0qik7QWAodG0h5TARUsmhnBfITrBTka81h+6JIN/ZTLq+abC
katI8VltU56K8HeXEcYTXuPWN3pjJAFlVphLPgvtxp5EwsFvykK6iebSKAKFfe5w
dCxSTuL/G/T+PAQNMXxAGS5MiheRFHu7I70ES9ddRuH4qGkVVutnjY9skajcX2+a
w9m1s/fFovci/bOaYCfQqfYlWXLHnOS+ewl9kGOPC3re6l3byY4+Lr19cmrDlNE+
qp3igws7FnoOJOjETY5D5sBvd+AikZcBk9uvVhWubPpJ5uLHQJactj/UigJtfGsJ
rKCpwX7oN30ZeSRf3vjCDUqO6HKKHih9dsPhQ8ikjswfqqTuMU0oPoNL+9PSaZi7
k9X0qcTpsC0M+Da+K2egehus507NFwpLLnMAWftLhzQoIxdt4/YHNEf0QT7P2L5F
5zZwImCS
=zVl4
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `ccd43c0d933c3fca4055f6f6006fddbebd4ec81c67d0810961187dc2dab9838d  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108ccd43c0d933c3fca4055f6f6006fddbebd4ec81c67d0810961187dc2dab9838df010a8cfbfabc9723aad4d8a0898801437dd08fff01042349393c8cde4166cc32202016813fb08f1045d3af328f008876d45244facee920083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010924f7e5552591120bb875304f256af3908f1045d3af328f008bf56f071bb6cd6b20083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010f3623c18aac31516d74dc3c420aa755a08f020a3e811bcdd75d6b89ff676656feed17d04f721b34e73e108e5e006f7dbd0af8708f1045d3af327f008a71bf20592b123470083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010fea51a72035ac10362d02724415e7cb908f1200ba482d6f930a93b3dd301a121071d726fad3432db019b82d21cc563eb1f773608f1045d3af327f0083bb1f539f1f1a7650083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

MarcoFalke commented:
I am seeing segmentation faults when shutting down during init:

```
==4778== Thread 9 bitcoin-shutoff:
==4778== Invalid read of size 4
==4778==    at 0x4953BC: FlushBlockFile(bool) (validation.cpp:1548)
==4778==    by 0x496B76: CChainState::FlushStateToDisk(CChainParams const&, CValidationState&, FlushStateMode, int) (validation.cpp:2061)
==4778==    by 0x497AB5: CChainState::ForceFlushStateToDisk() (validation.cpp:2115)
==4778==    by 0x32AE19: Shutdown(InitInterfaces&) (init.cpp:239)
==4778==    by 0x223F6F: BitcoinCore::shutdown() (bitcoin.cpp:161)
==4778==    by 0x58F4BF9: QObject::event(QEvent*) (in /usr/lib64/libQt5Core.so.5.12.4)
==4778==    by 0x4BFCAF5: QApplicationPrivate::notify_helper(QObject*, QEvent*) (in /usr/lib64/libQt5Widgets.so.5.12.4)
==4778==    by 0x4C05E7F: QApplication::notify(QObject*, QEvent*) (in /usr/lib64/libQt5Widgets.so.5.12.4)
==4778==    by 0x58C9AE7: QCoreApplication::notifyInternal2(QObject*, QEvent*) (in /usr/lib64/libQt5Core.so.5.12.4)
==4778==    by 0x58CCA92: QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) (in /usr/lib64/libQt5Core.so.5.12.4)
==4778==    by 0x591EE46: ??? (in /usr/lib64/libQt5Core.so.5.12.4)
==4778==    by 0x6F2EEDC: g_main_context_dispatch (in /usr/lib64/libglib-2.0.so.0.6000.5)
==4778==  Address 0x4 is not stack'd, malloc'd or (recently) free'd
==4778== 
==4778== 
==4778== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==4778==  Access not within mapped region at address 0x4
==4778==    at 0x4953BC: FlushBlockFile(bool) (validation.cpp:1548)
==4778==    by 0x496B76: CChainState::FlushStateToDisk(CChainParams const&, CValidationState&, FlushStateMode, int) (validation.cpp:2061)
==4778==    by 0x497AB5: CChainState::ForceFlushStateToDisk() (validation.cpp:2115)
==4778==    by 0x32AE19: Shutdown(InitInterfaces&) (init.cpp:239)
==4778==    by 0x223F6F: BitcoinCore::shutdown() (bitcoin.cpp:161)
==4778==    by 0x58F4BF9: QObject::event(QEvent*) (in /usr/lib64/libQt5Core.so.5.12.4)
==4778==    by 0x4BFCAF5: QApplicationPrivate::notify_helper(QObject*, QEvent*) (in /usr/lib64/libQt5Widgets.so.5.12.4)
==4778==    by 0x4C05E7F: QApplication::notify(QObject*, QEvent*) (in /usr/lib64/libQt5Widgets.so.5.12.4)
==4778==    by 0x58C9AE7: QCoreApplication::notifyInternal2(QObject*, QEvent*) (in /usr/lib64/libQt5Core.so.5.12.4)
==4778==    by 0x58CCA92: QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) (in /usr/lib64/libQt5Core.so.5.12.4)
==4778==    by 0x591EE46: ??? (in /usr/lib64/libQt5Core.so.5.12.4)
==4778==    by 0x6F2EEDC: g_main_context_dispatch (in /usr/lib64/libglib-2.0.so.0.6000.5)

jamesob commented:
Good catches, @MarcoFalke! I've [pushed fixes](https://github.com/bitcoin/bitcoin/compare/c860b44ce33f424de7f0fe5f96efa65c58bdb531..8cd527d72442e76cfc6dc848eb01b52f2a8f2d4c), including https://github.com/bitcoin/bitcoin/pull/16443/commits/e4b3247d4a67f248ff13bf43c378337ee993555b to solve the init bug you found. There are a few ways of addressing that one but I think an early exit from `FlushStateToDisk()` is the cleanest since it doesn't involve partial construction of `CoinsViews` or introducing more have-I-initialized-yet state, but let me know if you think there's a better way.

ryanofsky commented:
> In commit [53c8d59](https://github.com/bitcoin/bitcoin/commit/53c8d59d182684f33f944e3513b56d3be4f557d2) refactor: have CCoins* data managed under CChainState
> 
> ```
>    - The initialization order in init.cpp is changed slightly: the coinsdb (leveldb)
>       creation is moved to be earlier than block index and genesis block loading,
>       though this shouldn't have any noticeable effect.
> ```
> 
> Could you elaborate a bit more why it is save to flush the chainstate when shutdown is requested during the init sequence?

I'm also curious about the answer to this. Would be good to say what's motivating the move in the comment. I guess I also had a similar question previously:  https://github.com/bitcoin/bitcoin/pull/15606#discussion_r271901526

jamesob commented:
@ryanofsky @MarcoFalke good points. I've abandoned trying to reorder the init.cpp stuff and now have separated the construction of `CChainState` from `CoinsViews` to preserve the existing order.

ryanofsky reviewed (approved):
utACK 4e307e5f5d676d01a94e1b645dcb0ee8574faf77. Only changes since last review are adding and using little InitCoinsViews / HasCoinsViews helpers so the init order doesn't have to change, and reverting the new FlushStateToDisk check which is no longer needed because the init order is not changing. Also, Marco's name is now added to the credits, but I can live with that.

```diff
-Thanks to Russell Yanofsky for helpful feedback.
+Thanks to Russell Yanofsky and Marco Falke for helpful feedback.
```

ariard commented on src/validation.cpp:
```diff
@@ -314,7 +314,8 @@ bool CheckSequenceLocks(const CTxMemPool& pool, const CTransaction& tx, int flag
 
-static void LimitMempoolSize(CTxMemPool& pool, size_t limit, unsigned long age) EXCLUSIVE_LOCKS_REQUIRED(pool.cs)
+static void LimitMempoolSize(CTxMemPool& pool, size_t limit, unsigned long age)
+    EXCLUSIVE_LOCKS_REQUIRED(pool.cs, ::cs_main)
```
Based on above discussion on lock annotations, you may use also `AssertLockHeld` here and in `UpdateTip`
(also `AcceptToMemoryPoolWithTime`, `DisconnectTip`, `ConnectTip`)

ariard reviewed (approved):
ACK 4e307e5 (reviewed changes since last one and ran tests)

MarcoFalke commented:
ACK 4e307e5f5d676d01a94e1b645dcb0ee8574faf77 (only checked that my name is in the credits, did not look at the changes)

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 4e307e5f5d676d01a94e1b645dcb0ee8574faf77 (only checked that my name is in the credits, did not look at the changes)
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhusgwAu958HvbN9frbzwlJBJbL9+PABYM8kmB/+hArsWHs5J9vxPgDIyeGCsGN
fma8zMP/2Ru+YNBkr8fI8Dw24W7yrr+HliAzwV5czY9RFH+BJa4lON9lxs8Ni+07
cMwQe0I1mYYerOK8GEwmFqluyUpRx9EI5FF0gFsrmz4zKJI5KTGYv1XZ2ILLV4we
Zf3c58H0xLxi2RUywUIqNvYSYqsuJS1hf9bsvVnyzeJh0Uq84o74qnC9Y7Hx3Uy4
4gd4IzxUpiAeToE8KNpPfT2T+DmHNiuREqqmezUOdWkI1DN/RYY/jDORr2CysTvt
7mrrSpmNxbfTiboHZZvpxlFb94zQfxW08n9ex76z1N6aknbu7sTVW62Z+yQKU0LR
Nmcmxrg7gQ5uJU3B/cciM7TpJtHsOWHyUu1gDHGGBluK9kqTBt6jVU+XCwuJAUrU
hOgzIbf4qqIY3Bb5rq04RN/QBTDRZr43by8cX1HmyXNYXWaPQ2ZAVTdIgwYIPkjl
LLr6U7+n
=Lsbq
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `055fc53ddab45b0500c4174e6690d5cf015f456b1806475966f66a2d74e0de0c  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108055fc53ddab45b0500c4174e6690d5cf015f456b1806475966f66a2d74e0de0cf010d158ba28ece332831d003740848d68bf08fff01014aea214a20e5147cbd9f399f6c8be5a08f1045d4184b9f0084f85b66eb6daf0360083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010540308af02f41071ae1fac1ea77b1c5f08f120aafef8c076d8513ae3f2b4eaf60d9e370b350b98012c82ee32941028d294f84808f1045d4184b8f00879021ce0c570ec7b0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010ea5857e03a295027d59f78b031bd7e3b08f120f0f0a806c9888989c308434166ad7e3b02b548b8c5c8fa7f6a323c6a563cbde308f1045d4184b8f00811f3e36f870a14100083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010fc2a7da34af640f5329a354cbea3422608f120d0d2313982a98175476bfc5fee63a3d6d853277581d63639afb92e6ed28a35e508f1045d4184b8f008688f3dfa5a9a9a5d0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

MarcoFalke commented on src/init.cpp:
```diff
@@ -1520,33 +1519,34 @@ bool AppInitMain(InitInterfaces& interfaces)
                 }
 
-                // The on-disk coinsdb is now in a good state, create the cache
-                pcoinsTip.reset(new CCoinsViewCache(pcoinscatcher.get()));
```
Again, would be nice if you explained why it is safe to flush the db when we were unable to replay the blocks or upgrading the db, ... $any_other_error

jamesob replied:
Good point. Instead of changing any of the `init` ordering and explaining with a comment, I've made further changes that split up initialization of the coinsdb and coinscache.

jamesob commented:
I've pushed [changes](https://github.com/bitcoin/bitcoin/compare/4e307e5f5d676d01a94e1b645dcb0ee8574faf77..72258a94731d662122880eedc9b0e278a09d455c) ([`au.coins-under-chainstate.9`](https://github.com/jamesob/bitcoin/tree/au.coins-under-chainstate.9) -> [`au.coins-under-chainstate.10`](https://github.com/jamesob/bitcoin/tree/au.coins-under-chainstate.10)) that allow separate initialization of the coinsdb and coinscache so that we can follow existing `init` semantics without having to think about whether flushing to disk after some kind of database initialization problem is safe or not. I then rebased onto master.

ryanofsky reviewed (approved):
utACK cdf6afbbf1e15710ebbd26c02ccbd834db1431de. Changes since last review: rebase, locking cs_main while flushing during shutdown in third commit, adding new asserts to check initialization order, renaming some methods, and splitting coinsdb / coinscache initialization as described in second commit.

Sjors reviewed (approved):
Code review ACK cdf6afbbf1e15710ebbd26c02ccbd834db1431de. I also lightly stress tested it by  quitting QT while rolling back and forward (with `--enable-debug`).

Maybe @promag (given his work on #15997) can comment on the last commit that adds lock annotations.

TheBlueMatt commented:
Does it make sense to go ahead and do some kind of ::ChainstateActive().CoinsTip(I_WANT_THE_FULLY_VALIDATED_ONE_ONLY vs IM_OK_WITH_ASSUMEVALID) distinction in this pr?

MarcoFalke commented:
Why would you need such a distinction at the coins tip level? I'd rather have a ::ChainstateBackground() global.

jamesob commented:
@TheBlueMatt no - right now there's no notion of an assumeutxo-valid chainstate. Later on when we want to acknowledge the possibility of multiple chainstates we'll do so through the [`ChainstateManager`](https://github.com/bitcoin/bitcoin/commit/bf4a50371448d4a995ccd8e0cf299986b96491fe#diff-349fbb003d5ae550a2e8fa658e475880R806) interface (as @Sjors noted).

ariard commented on src/init.cpp:
```diff
@@ -235,8 +235,11 @@ void Shutdown(InitInterfaces& interfaces)
-    if (g_chainstate && g_chainstate->CanFlushToDisk()) {
-        g_chainstate->ForceFlushStateToDisk();
+    {
+        LOCK(cs_main);
```
Don't know about this lock, there is no thread which reset `g_chainstate`, and if we shutdown before we init `g_chainstate` we should be still single-threaded. Anyway, taking a lock at shutdown shouldn't hurt

jamesob replied:
Yeah, only added this to satisfy the new lock annotations on `CanFlushToDisk()`.

ariard commented on src/validation.h:
```diff
@@ -505,8 +505,40 @@ class BlockManager {
+    //! can fit per the dbcache setting.
+    std::unique_ptr<CCoinsViewCache> m_cacheview;
+
+    //! This constructor initializes CCoinsViewDB and CCoinsViewErrorCatcher instances, but it
```
nit: if you rebase, you may add a reference to this in commit message, at would ease understanding of changed since last time :)

jamesob replied:
Updated the message, thanks.

ariard reviewed (approved):
ACK cdf6afb, reviewed changes since last time and run tests with `--enable-debug` for lock checking.

MarcoFalke commented on src/init.cpp:
```diff
@@ -1520,33 +1522,38 @@ bool AppInitMain(InitInterfaces& interfaces)
 
-                // The on-disk coinsdb is now in a good state, create the cache
-                pcoinsTip.reset(new CCoinsViewCache(pcoinscatcher.get()));
+                // The on-disk coinsdb is now in a good state, create the cache.
```
in commit 087bdf7190 refactor: have CCoins* data managed under CChainState


Why is this comment changed? None of the other comments have trailing dots

MarcoFalke commented on src/validation.h:
```diff
@@ -505,6 +505,41 @@ class BlockManager {
+    std::unique_ptr<CCoinsViewDB> m_dbview GUARDED_BY(cs_main);
+
+    //! This view wraps access to the leveldb instance and handles read errors gracefully.
+    std::unique_ptr<CCoinsViewErrorCatcher> m_catcherview GUARDED_BY(cs_main);
```
In commit 087bdf7190 refactor: have CCoins* data managed under CChainState


Why is this a pointer?

MarcoFalke commented on src/validation.h:
```diff
@@ -505,6 +505,41 @@ class BlockManager {
+public:
+    //! The lowest level of the CoinsViews cache hierarchy sits in a leveldb database on disk.
+    //! All unspent coins reside in this store.
+    std::unique_ptr<CCoinsViewDB> m_dbview GUARDED_BY(cs_main);
```
Same

MarcoFalke commented on src/validation.h:
```diff
@@ -553,19 +588,71 @@ class CChainState {
+    CCoinsViewCache& CoinsTip() EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+    {
+        assert(m_coins_views->m_cacheview);
+        return *m_coins_views->m_cacheview.get();
```
in commit 087bdf7190 refactor: have CCoins* data managed under CChainState

No need for `.get()`;

(Same below)

MarcoFalke commented on src/validation.cpp:
```diff
@@ -1053,7 +1054,7 @@ CoinsViews::CoinsViews(
                         m_catcherview(MakeUnique<CCoinsViewErrorCatcher>(m_dbview.get())) {}
 
-void CoinsViews::InitCache()
+void CoinsViews::InitCache() EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
```
in commit cdf6afbbf1

Please don't add lock annotations in the cpp file. They will overwrite the ones in the header file

jamesob replied:
Oops, stupid mistake.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -1073,7 +1074,7 @@ void CChainState::InitCoinsDB(
 }
 
-void CChainState::InitCoinsCache()
+void CChainState::InitCoinsCache() EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
```
Same

MarcoFalke commented:
partial ACK 087bdf7190

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 087bdf7190
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiSqQv+PyvXoFWuFIO8A5dCEelRFbODvV6DCXgYAdsWbKLeJQL9djcqnVE1xNHS
XI90uua6p2kjXYOxjDn9l4r3mGizgYJkbiJmjwJef++U8sjGAPhhlxnPvS3X0UMU
83qLml/tGeIWdW/ynlYq/wSQlADMyKfIA6QYyTHlNAN0IPKsVnPtN/RhAJmuyr+z
OvegeJQ+DFJB+Kc7z5SNx0I/c29K0FJhISZkTpGAaeA/fIfyg2p/xIcLnYbLzGnd
m4dsNiklqNJRln3I0qdctuBpX3NCcVOOPfAPW5bJvIGtVbSObY7d09eNqOiS98h7
HRCOLN/MSccoUpXSSb+IdAnVHG+333cKnOSbMzdmsRc6Jtytgc3qTzFtv9MWjoty
GzmLEIchjLhDPOFCzYDb38kkIRNQgvCg+ldfQ7TBuUz76oXGz/kSNuve+V6gzL61
tJOyiW/JTCXJEOEAOTzPfDpybNqIcfkca0bfwOhRnBgC7iJ3U/fRCW2eVuSWryFS
Knp5/0Pk
=w+eT
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `0a4a6fbe53dfa1d0e95a1dcea6c07b8d89dd49766c5459918ae9ed368258510d  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401080a4a6fbe53dfa1d0e95a1dcea6c07b8d89dd49766c5459918ae9ed368258510df01024f784fb5943a13ccc3ce19ed07de35508fff0100e12f85be96e31f1ac5f2c043bf94e6c08f020747de850e0345124135c01dd66c19aa48ceef6699c03f0a12fb65f62f78c350c08f1045d5562a1f0086c658ef110ffe9b70083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0101e7d90f2af1416b5630316db7dbb5f6f08f12098c6d97a38b2c9039f6853bcdeb9b7a5a4c886c9aa1e303204566cfb1513b8f208f1045d5562a1f0089859447721714b960083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff01043f08667f4e1dee2e70547137ddcb5d008f1045d5562a1f00810dc43bf00dea85c0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010bdd248b4d6a2f8e999ae7d29147e8d8408f1045d5562a1f00885dcc96322a9ab5c0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

jamesob commented:
[`au.coins-under-chainstate.10`](https://github.com/jamesob/bitcoin/tree/au.coins-under-chainstate.10) -> [`au.coins-under-chainstate.12`](https://github.com/jamesob/bitcoin/tree/au.coins-under-chainstate.12) 
- [(changes `10` -> `11`)](https://github.com/bitcoin/bitcoin/compare/cdf6afbbf1e15710ebbd26c02ccbd834db1431de..c41bcbd3ef6a61335c4e144633bc3ce6e52abcb7) 
- [(changes `11` -> `12`)](https://github.com/bitcoin/bitcoin/compare/c41bcbd3ef6a61335c4e144633bc3ce6e52abcb7..582d2cd74754d6b9a2394616a9c82a89d2d71976) 

Thanks for finding that stuff @MarcoFalke; fixes pushed.

Sjors commented:
reACK 582d2cd74754d6b9a2394616a9c82a89d2d71976

MarcoFalke commented:
ACK 582d2cd747

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 582d2cd747
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUinvAv+MV+6LkNCYGqrYG1HI3cm/fQYw8DmpC0nsUHkaEDHCfaqTvAimUoXxveq
zIb2gsCgbyujAM+zCTLlUBdE1v4stJ1fPMq57afluQNpEXrRSYttdy33YAI5D4Dy
QJ/Ny9kHTdTMukYe8eVnDqNWgwIgjfPkzFrhOw3vlN45e6aVQl7WX+yvaaB5v1/A
fg8eZ93c8ey4X0P6hDaT2jdVNNWidlnXZ2+bbZOqEKUfI0boJtQrcp4oxUkJ40kf
A7kFuwBvKyxUs8HWLeTUW4P39RHBSuhjewaHuBVneJpPOw7I3tYcOUtZS7Rv8CJC
ZZnj+GVLnLyAz0HMD/KCuLGpFV28oxgHdMfqRkNfGvFJFkAsEfyuFjHjWBe9DT9A
owlq2UGeBlMznMC2UK+GIX+5TIKDBShgoXbblu9x8x7R+s9iSRHmEiSEGtOst6qP
OU6mr2hOTB3qg1wywWHNbxfqpJiuP6TknWOFm7G5GnJcSrKUfO42RII0Q2E/oVQ2
QxgRxcan
=/V5U
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `8c58c7b1c21f3568020ef04ac438a92aba830446a5172e171b37e02c744d64e0  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401088c58c7b1c21f3568020ef04ac438a92aba830446a5172e171b37e02c744d64e0f010b4c17127a67052ec2aef375c8cd5726d08fff0101905520377e42f2a64b47eb0b52b2cf908f1045d558c5ef008227ccf2a832775380083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0106a21d1bef196775e3d59521f3061c6b508f120af902da0d19648c9ae2fbebaf6503d0e4cdfff8776bf40cf77876580dfaf2c1f08f1045d558c5ef008a2a96f23b6aaf1490083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010c1bc7106cf11008369f9f553beea6bfd08f120f6c0a6ea479510b913cd49f9cd33686c334e320c124298a59e0eec2841553e0e08f1045d558c5ef0082853382e7d12bbb60083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010d003a7c24afe9f210f7992f1e7a9448308f1203d9dea539db4393245a66d454a49015bdcedc01ea4c65701dbb4c33f7f032cf008f1045d558c5ef008de95b42763fedd9e0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>
